### PR TITLE
test: Add test for "needCheckForUpdates"

### PR DESCRIPTION
**Pull request by tanishiking**
https://github.com/scalameta/metals-vscode/issues/1144

This commit adds a test for `needCheckForUpdates`.

For adding unit tests, this commit abstract away the logic that needs
access to vscode APIs, becuase we need to run an integration test for
accessing vscode APIs.
see: https://github.com/microsoft/vscode-wordcount/issues/5#issuecomment-171050759
https://github.com/microsoft/vscode/issues/82471#issuecomment-564731132

### CODE REVIEW

**Comment by tanishiking** [comment]
Copied from https://github.com/tanishiking/metals-vscode/pull/82

> [kpodsiad](https://github.com/kpodsiad) commented [7 hours ago](https://github.com/tanishiking/metals-vscode/pull/82#issuecomment-1234763293)
> One problem with assert is that it doesn't provide meaningful error messages. I was thinking about using [earl](https://earljs.dev/docs/), wdyt?

Yeah, adding an assertion library is good 👍 
However, in that case, I would choose [chai](https://www.chaijs.com/) as its tried and tested solution (or I would pick `jest`  instead of `mocha + earljs` as all-in-one test framework, but I personally prefer `mocha + chai + sinon` for testing node project).

File: src/test/unit/checkForUpdate.test.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    const repo = new MockRepo(prevVersion, today);
+    const spy = sinon.spy(repo, "saveLastUpdated");
+    const actual = await needCheckForUpdates(
+      currentVersion,
+      today,
+      ConfigurationTarget.Global,
+      repo
+    );
+    expect(actual).false;
+    expect(spy.getCall(0).args).to.eql([
```
Comment (tanishiking) [inline src/test/unit/checkForUpdate.test.ts:60]: we can use `spy.calledWith(...)`, but this way will show better error message if test fail (while `calledWith` shows true or false)

**Comment by kpodsiad** [comment]
> Copied from [tanishiking#82](https://github.com/tanishiking/metals-vscode/pull/82)
> 
> > [kpodsiad](https://github.com/kpodsiad) commented [7 hours ago](https://github.com/tanishiking/metals-vscode/pull/82#issuecomment-1234763293)
> > One problem with assert is that it doesn't provide meaningful error messages. I was thinking about using [earl](https://earljs.dev/docs/), wdyt?
> 
> Yeah, adding an assertion library is good 👍 However, in that case, I would choose [chai](https://www.chaijs.com/) as its tried and tested solution (or I would pick `jest` instead of `mocha + earljs` as all-in-one test framework, but I personally prefer `mocha + chai + sinon` for testing node project).

Actually I was wrong. `assert` is good enough, it has meaningful errors and it doesn't require additional dependency. I intentionally switched jest to mocha in this repository simple because it's [much smaller](https://bundlephobia.com/package/mocha@10.0.0) than [jest](https://bundlephobia.com/package/jest@29.0.1) and it doesn't bring so many dependencies (I recall some problems with them)

**Comment by tanishiking** [comment]
> I intentionally switched jest to mocha in this repository simple because it's much smaller than jest and it doesn't bring so many dependencies

Yeah, I also think mocha chai sinon would be noice in this repository (than jest) :+1:
You don't want to add chai as a dependency? I prefer to use chai because I like BDD style testing and it's small library.

File: src/repository/CheckForUpdateRepo.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { ExtensionContext, Memento } from "vscode";
+import { ConfigurationTarget } from "../ConfigurationTarget";
+
+type LastUpdated = {
```
Comment (kpodsiad) [inline src/repository/CheckForUpdateRepo.ts:4]: I think it's more common to use interface for such data types

File: src/repository/CheckForUpdateRepo.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+export interface CheckForUpdateRepo {
+  getLastUpdated(target: ConfigurationTarget): LastUpdated;
+  saveLastUpdated(
+    serverVersion: string,
+    lastUpdatedAt: string,
+    target: ConfigurationTarget
+  ): void;
+}
+
+export class DefaultCheckForUpdateRepo implements CheckForUpdateRepo {
```
Comment (kpodsiad) [inline src/repository/CheckForUpdateRepo.ts:18]: SO it's not a problem that interface is located in the same file which has imports from `vscode` package as long as it doesn't use them, nice. Maybe I'll rewrite/write some unit tests for test explorer then :D

**Review by kpodsiad — APPROVED** [review]
> You don't want to add chai as a dependency? I prefer to use chai because I like BDD style testing and it's small library.

It's ok, I just meant that I forgot that node's assert also provides good errors ;) 

I think we're missing one important think (correct me if I' worng) - do we have some docs how to run single test file, single test from suite? That would be nice to have.
TBH, even `yarn test` might be useful for people like don't do js day to day.

File: src/repository/CheckForUpdateRepo.ts (hunk 2 above)
Comment (tanishiking) [reply src/repository/CheckForUpdateRepo.ts:4]: I don't have a strong opinion, but I prefer to use `type` over `interface` for the data types that shouldn't be extended (by inheritance or declaration merging).

**Comment by tanishiking** [comment]
Thanks! Added docs on how to run tests https://github.com/scalameta/metals-vscode/pull/1145/commits/b70d5b3101f37be5e1724974d436beb2a9af78cf
merging after CIs pass

**Pull request merged**